### ISSUE TITLE
New CLI argument to disable config argument type inferencing.

### ISSUE THREAD

**Issue by haje01** [snakemake/snakemake#680]
I have a Snakemake based ETL process which takes a config parameter for date to process. It runs like this:

    snakemake --config date=2020-10-09 -j

In this case, the type of date parameter is `string`, and the value is '2020-10-09'. So, I implemented the code for that.

But with following command:

    snakemake --config date=2020-10-10 -j

My code raises an error. Because, now the type of date parameter is `int`, and the value is 2000. I guess some kind of smart implicit type inferencing and evaluation is working here. 

I think explicit is better than implicit. But, you may have reasons. So, I suggest simple extra command line argument to disable type inferencing:

    snakemake --config date=2020-10-10 --disable-config-type-inferencing -j

I would like to hear you opinions. 

Thank you.

**Comment by nhartwic** [comment]
I agree that passing "config" values through the CLI sucks. I don't think an option to disable type inference really solves anything though. The real problem is that config values are passed using an unspecified and ad hoc format. The sollution to this is to switch to a widely used format that snakemake already uses and that can work at the command line: json. Your command would be...

```
snakemake --config ' {"date": "2020-10-10"} ' -j
```

No more ambiguity. Works super cleanly with configfiles. Only drawback is that json is a bit ugly and needs to be quoted properly

**Comment by nhartwic** [comment]
I can probably put together a pull request to implement this, but it seems like a large change to the CLI and I'm hesitant to do so without input from other, more experienced snakemake contributors.

Does anyone have any thoughts about switching to json for passing CLI config options? Was anyone around when the current system was being implemented? Since it seems like using json is the obvious thing to do, I have to ask if there is a reason this wasn't done before?

**Comment by dariober** [comment]
@nhartwic 

```
snakemake --config ' {"date": "2020-10-10"} ' -j
```

This doesn't work for me on snakemake 5.14.0:

```
snakemake --config ' {"date": "2020-10-10"} ' -j
Invalid config definition: Config entries have to be defined as name=value pairs.
```

(I can't/don't want to upgrade snakemake because of this issue https://github.com/snakemake/snakemake/issues/575)

**Comment by nhartwic** [comment]
You can't use a json string to pass config values via the 'config' option in any existing version of snakemake. This issue is about an enhancement proposal, a change to current functionality.

In the meantime, you will probably need to write your json string to a file and then pass that file to snakemake using the 'configfile' option. 

@dariober

**Comment by bingxiao** [comment]
Hi Authors,

I've experienced the same problem. It seems for each `key=val` pair, the config parser parses `val` using `int, float, eval, str` in turn. The consequences are `x=1+2` will be eval-ed to `x=3`, and `label=0x10` to `label=16`, etc. Even though sometimes users may want to use the python syntax convenience on the command line, there are other cases they truly need string tags like `1+2` and `2+1`.

A small change to the code was mentioned above. The reason for it is to keep information intact when flowing from cmd to snakefile. Glad to know if the quickfix is helpful.

Thanks.

https://github.com/snakemake/snakemake/blob/367607f0929f8cc3b4dee0ea07e75c5018a691b1/snakemake/__init__.py#L869-L874

**Comment by nhartwic** [comment]
@bingxiao What quickfix are you referring to? What was the former behavior and what behavior are you suggesting moving too?

Frankly, anything with the power of "eval" behind it seems like a terrible idea to me. My principle concern here is that the snakemake workflow shouldn't have to know or care whether or not a value was passed via configfile or via config. The only way to make that work is to make "config" and "configfile" equally powerful. Any object that can be specified with one needs to be specifiable with the other. "eval" can create all kinds of objects, example: sets, dataframes. Json and Yaml can't. Allowing users to pass different kinds of values via configfile and config is going to cause bugs. I'd much prefer an equally powerful langugage to having to rely on developers to not shoot themselves in the foot. 

Is there any real argument against eliminating the add hoc config language which isn't specified anywhere and just switching to json?

**Comment by bingxiao** [comment]
@nhartwic Thanks for your consideration. Anyway, I second your argument with the goal to moving cmd config method to json.

The reason for me to use the cmd config is I need to loop a long paired list of tags, like `--config label_1={k}_1 sample=sample_{k}`, thinking it could be easily done by generating a cmd list in bash. I previously worked around this issue by adding a line after parsing checking that `1_1` is still `1_1`, and it should also work by removing `eval` parser completely. Sanitising the current parser could still be a sensible option before the json option is ready in cmd and being widely adopted.

Didn't send a PR because not sure the modification was sensible enough.

**Comment by johanneskoester** [comment]
Thanks a lot for pointing me to this issue. I have now merged PR #709, which simply interprets complex values via the YAML parser. This makes most sense to me, as it then behaves like putting them in the config file directly.
